[benchmark] Fix benchmarks for Set operations

lorentey opened the pull request:
- Don’t use a random number generator. (Hashing already scrambles elements.)
- Add benchmarks for `Set.subtracting`.
- Add new benchmarks for 0%, 25%, 50% and 100% overlap between sets. (The latter two only for `Set<Int>`.)
- Also keep original benchmarks where feasible. (These only tested disjunct sets, so they were relatively useless.)

lorentey committed:
[benchmark] Fix benchmarks for Set operations

- Don’t use a random number generator
- Ensure there is a 25% overlap between sets. Original benchmarks tested non-overlapping sets only.
- Change names; results aren’t directly comparable.

lorentey commented:
@swift-ci please smoke test

lorentey commented:
@swift-ci please benchmark

airspeedswift commented on benchmark/single-source/SetTests.swift:
```diff
@@ -13,106 +13,69 @@
-  BenchmarkInfo(name: "SetIsSubsetOf_OfObjects", runFunction: run_SetIsSubsetOf_OfObjects, tags: [.validation, .api, .Set]),
-  BenchmarkInfo(name: "SetUnion", runFunction: run_SetUnion, tags: [.validation, .api, .Set]),
-  BenchmarkInfo(name: "SetUnion_OfObjects", runFunction: run_SetUnion_OfObjects, tags: [.validation, .api, .Set]),
+  BenchmarkInfo(name: "SetExclusiveOr2", runFunction: run_SetExclusiveOr2, tags: [.validation, .api, .Set]),
```
The `2` so we don't track this as a change in performance of this benchmark? @eeckstein is this important or should we not bother?

eeckstein replied:
It is important. Renaming avoids showing this changes as false compiler improvements/regressions.

airspeedswift reviewed (approved):
Nice

palimondo commented:
Since we are renaming these, can we take also lower the base workloads on these benchmarks to run under 1ms?

lorentey committed:
[benchmark] SetTests: Better parametrization, separate setup

lorentey committed:
[benchmark] SetTests: Re-add original tests

The code isn’t the same, but comparisons with previous releases will still be useful (as long as timings match with the original code).

lorentey commented:
Welp, serves me right for not going the whole way. 

The latest commit adds benchmarks with names and parameters matching the original benchmarks. (But without setup costs etc.) If they come out close to the originals, I think there's still value in keeping them, so that we have an easy way to compare releases.

(Swift 4.2 and 5 are both extensively changing the internals of hashing as well as Set and Dictionary itself; it makes sense not to break old benchmarks if we can help it -- even if they are of questionable practical value.)

lorentey commented:
I need to try the shiny new thing!

@swift-ci please smoke benchmark staging

lorentey commented:
@swift-ci please smoke test

lorentey commented:
@swift-ci please smoke test

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -12,69 +12,115 @@
+let size = 400
+let overlap = 100
+
+let setAB = Set(0 ..< size)
```
I was drawing number lines like the dummy that I am… a little bit of spelling out would help me grok the correctness of this much sooner. What do you think?
````swift
let setAB = Set(0 ..< size)                            // 0...399
let setCD = Set(size ..< 2 * size)                     // 400...799
let setBC = Set(size - overlap ..< 2 * size - overlap) // 300...699
let setB = Set(size - overlap ..< size)                // 300...399
````

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -12,69 +12,115 @@
+    setUpFunction: { blackHole([setAB, setBC]) }),
+  BenchmarkInfo(
+    name: "SetExclusiveOr2_OfObjects",
+    runFunction: { n in run_SetExclusiveOr_OfObjects(setOAB, setOBC, countAC, 10 * n) },
```
I'm skeptical that the same multiplier for objects as for `Int`s woudn't overshoot the healthy runtime. Can we aim for 1ms (1000 μs)?

lorentey replied:
The difference between `Int` and `Box<Int>` is relatively small; it's merely an extra indirection. It is important to remain able to keep track of the difference, though.

Capitalized names for function parameters is a silly tradition; I'm happy to be the barbarian who breaks it. ;-)

lorentey replied:
(Oh, beyond the indirection, Box<Int> also hashes a bit slower than Int since it's not forwarding the one-shot hashing implementation.)

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -12,107 +12,156 @@
+    tags: [.validation, .api, .Set],
+    setUpFunction: { blackHole([setAB, setBC]) }),
+  BenchmarkInfo(
+    name: "SetExclusiveOr_OfObjects",
```
Could you please drop the underscores from test names, at least from the new ones? I'd call them like `SetExclusiveOrBoxed`, if coming up with the name from scratch.

lorentey replied:
Sure, as long as we end up losing compatibility. But if some of the original names remain, I'd prefer to make the relationship between old/new tests obvious. (Naming consistency is far less important than utility in this case.)

palimondo replied:
@eeckstein cares a lot about continuity of perf-tracking… I might have a look at adding some `lnt` migration scripts for that if it remains a point of contention.

lorentey committed:
[benchmark] SetTests: Explain input sets

lorentey commented:
@swift-ci please smoke benchmark staging

lorentey commented:
@swift-ci please smoke benchmark staging

palimondo reviewed (approved):
LGTM (Legendary Great ™)*

`*`   pending rename of new benchmarks (old too, when continuity fails)

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -12,107 +12,156 @@
+  _ r: Bool,
+  _ n: Int) {
+  for _ in 0 ..< n {
+    let isSubset = a.isSubset(of: identity(b))
```
One more thing… could you document what's the purpose of wrapping `b` in `identity` here?

lorentey replied:
It should be a relatively common idiom in these benchmarks -- it's there to prevent the compiler from moving things out of the loop. `a` and `b` are constant through all iterations, and in theory, a sufficiently smart compiler could optimize some/all of it away. Adding an opaque function call with unknown effects (hopefully) defeats these optimizations.

(I don't think any optimizations would apply in these cases, but it's better to be safe than sorry.)

palimondo replied:
That explanation makes perfect sense. How about putting it in the source comment? That way it’s easier to point people unfamiliar with the idiom to this file as an example of best practice (link from docs, when we get to it).

lorentey committed:
[benchmark] SetTest: Add tests for more operations and more cases

- Isolate legacy tests. Add new tests for the same operations, with updated iteration counts and names.
- Rename new tests to follow a consistent naming scheme.
- Add tests for Set.subtracting.
- Add tests on integer sets with 50% and 100% overlap. (isSubset, intersection, union, symmetricDifference, subtracting)

lorentey committed:
[benchmark] SetTests: Move legacy benchmarks to end of list.

lorentey commented:
@swift-ci please smoke benchmark staging

lorentey commented:
@swift-ci please smoke benchmark staging

lorentey commented:
@swift-ci please smoke test

lorentey commented:
@swift-ci smoke test

lorentey commented:
Kindly
@swift-ci smoke test macOS platform

lorentey commented:
Having a variety of non-disjunct cases helps to keep the code honest -- otherwise we could end up optimizing for particular inputs and neglect/harm the others. 

E.g. we currently perform set subtractions by elementwise removal, which optimizes for the disjunct case with uniquely held storage, but it isn't very good when faced with COW copies or large overlaps.

palimondo commented on benchmark/single-source/SetTests.swift:
```diff
@@ -12,107 +12,288 @@
+  BenchmarkInfo(
+    name: "SetIsSubsetInt0",
+    runFunction: { n in run_SetIsSubsetInt(setAB, setCD, false, 5000 * n) },
+    tags: [.validation, .api, .Set],
```
Tags are same for all test here. How about extracting this to constant and DRY the declarations a bit?

`name: "Set…", tags: tags,`

palimondo commented:
Excellent work! So glad to see cleanup and rejuvenation of the old crusty tests!
